### PR TITLE
[HttpClient] exceptions carry response

### PR DESCRIPTION
**Pull request by antonch1989**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #30502
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/HttpClient/Exception/HttpExceptionTrait.php [hunk 1]
```diff
@@ -20,8 +20,12 @@
  */
 trait HttpExceptionTrait
 {
+    /** @var ResponseInterface */
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Exception/HttpExceptionTrait.php]: should be removed

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for starting this.
I think we should add the method in an interface in contracts.
If we add it to the base `ExceptionInterface`, it should be nullable, because the response object might not be instantiated yet when this is thrown.
We could alternatively add a new base `HttpExceptionInterface` that would add this method to `Server/Client/RedirectionExceptionInterface` - that's mean the method is not on `TransportExceptionInterface`.
Any preference?

File: src/Symfony/Component/HttpClient/Exception/HttpExceptionTrait.php (hunk 1 above)
Comment (antonch1989) [reply src/Symfony/Component/HttpClient/Exception/HttpExceptionTrait.php]: the solution with a new interface looks cleaner, I'll do this way if there are no objections

File: src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Symfony\Contracts\HttpClient\Exception;
+
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+interface HttpExceptionInterface
```
Comment (linaori) [inline src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php]: What do you think of `HttpResponseExceptionInterface`? The current name doesn't imply it has anything to do with a response, while the getter implies there's _always_ a response

Comment (antonch1989) [reply src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php]: In my opinion the current naming is fine, let's wait for some other comments

File: src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace Symfony\Contracts\HttpClient\Exception;
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php]: this file is missing the license header

File: src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php]: I'm fine with the name `HttpExceptionInterface` personnaly, but the interface should extend `ExceptionInterface`

File: src/Symfony/Contracts/HttpClient/Exception/RedirectionExceptionInterface.php [hunk 4]
```diff
@@ -18,6 +18,6 @@
  *
  * @experimental in 1.1
  */
-interface RedirectionExceptionInterface extends ExceptionInterface
+interface RedirectionExceptionInterface extends ExceptionInterface, HttpExceptionInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Exception/RedirectionExceptionInterface.php]: `interface RedirectionExceptionInterface extends HttpExceptionInterface`
please do the same `ClientExceptionInterface` and `ServerExceptionInterface`

File: src/Symfony/Contracts/HttpClient/Exception/ClientExceptionInterface.php [hunk 5]
```diff
@@ -18,6 +18,6 @@
  *
  * @experimental in 1.1
  */
-interface ClientExceptionInterface extends ExceptionInterface
+interface ClientExceptionInterface extends ExceptionInterface, HttpExceptionInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Exception/ClientExceptionInterface.php]: `ExceptionInterface` should be removed as it's now embeded in HttpExceptionInterface
same in the 2 other exception interfaces

File: src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php]: don't miss that one

File: src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Interface for getting response from an exception
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Exception/HttpExceptionInterface.php]: Base interface for HTTP-related exceptions.

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @antonch1989.

**Pull request merged**